src/sql/src/pure.rs

```diff
@@ -1475,3 +1482,239 @@ async fn compile_proto(
+        ) {
+            option.value = Some(WithOptionValue::Refresh(RefreshOptionValue::At(
+                RefreshAtOptionValue {
+                    time: mz_now_expr.clone(),
+                },
+            )));
+        }
+
+        // 2. If `REFRESH EVERY` doesn't have a `STARTING AT`, then add `STARTING AT mz_now()`.
+        if let Some(WithOptionValue::Refresh(RefreshOptionValue::Every(
+            RefreshEveryOptionValue { aligned_to, .. },
+        ))) = &mut option.value
+        {
+            if aligned_to.is_none() {
+                *aligned_to = Some(mz_now_expr.clone());
+            }
+        }
+
+        // 3. Substitute `mz_now()` with the timestamp chosen for the CREATE MATERIALIZED VIEW
+        // statement. (This has to happen after the above steps, which might introduce `mz_now()`.)
+        match &mut option.value {
+            Some(WithOptionValue::Refresh(RefreshOptionValue::At(RefreshAtOptionValue {
+                time,
+            }))) => {
+                let mut visitor = MzNowPurifierVisitor::new(mz_now, mz_timestamp_type.clone());
+                visitor.visit_expr_mut(time);
+                introduced_mz_timestamp |= visitor.introduced_mz_timestamp;
+            }
+            Some(WithOptionValue::Refresh(RefreshOptionValue::Every(
+                RefreshEveryOptionValue {
+                    interval: _,
+                    aligned_to: Some(aligned_to),
+                },
+            ))) => {
+                let mut visitor = MzNowPurifierVisitor::new(mz_now, mz_timestamp_type.clone());
+                visitor.visit_expr_mut(aligned_to);
+                introduced_mz_timestamp |= visitor.introduced_mz_timestamp;
+            }
+            _ => {}
+        }
+    }
+
+    // 4. If the user didn't give any REFRESH option, then default to ON COMMIT.
+    if !cmvs.with_options.iter().any(|o| {
+        matches!(
+            o,
+            MaterializedViewOption {
+                value: Some(WithOptionValue::Refresh(..)),
+                ..
+            }
+        )
+    }) {
+        cmvs.with_options.push(MaterializedViewOption {
+            name: MaterializedViewOptionName::Refresh,
+            value: Some(WithOptionValue::Refresh(RefreshOptionValue::OnCommit)),
+        })
+    }
+
+    // 5. Attend to `resolved_ids`: The purification might have
+    // - added references to `mz_timestamp`;
+    // - removed references to `mz_now`.
+    if introduced_mz_timestamp {
+        resolved_ids.0.insert(mz_timestamp_id);
+    }
+    // Even though we always remove `mz_now()` from the `with_options`, there might be `mz_now()`
+    // remaining in the main query expression of the MV, so let's visit the entire statement to look
+    // for `mz_now()` everywhere.
+    let mut visitor = ExprContainsTemporalVisitor::new();
+    visitor.visit_create_materialized_view_statement(cmvs);
+    if !visitor.contains_temporal {
+        resolved_ids.0.remove(&mz_now_id);
+    }
```

Comment:
Unfortunately, they can't be wrong in either direction. I had a bug during development where they were wrong, and I was hitting the `in-memory state of the catalog does not match its on-disk state` check in Testdrive tests. This is because the `resolved_ids` are stored in the in-memory state, and if a wrong set is being stored, then the set won't match the new set that we get when we reconstruct the in-memory catalog state from disk.

Reply:
What are these IDs used for?

Reply:
Ah, well, that seems plausible enough @ggevay. Let's leave the code as is then!

And @jkosh44 I was hoping you knew the answer to that question! I paged out the details long ago.

Reply:
I was hoping that @ggevay knew since this is a new function introduced in this PR. I took a look and these are later plumbed into planning, which uses these to check if any function in `mz_unsafe` is called. It's later plumbed into sequencing which uses them to determine required RBAC privileges and is then saves them in the catalog. The catalog assumes these are IDs that are directly referenced by the SQL statement.

Reply:
Yeah, @ggevay was forced to handle them here, but the resolved IDs were already created by the coordinator _before_ Gabor inserted this function, and then handled by planning/the coordinator _after_ this function, so there are some un/underdocumented guarantees that planning/sequencing require.

Reply:
I'm having trouble articulating exactly what the requirements are, but it's something like this.

When we receive a user DDL, we make the following transformation:
`user-ddl-sql -> (persisted-sql, resolved-ids)`. The `resolved_ids` input/output to this function is exactly those `resolved-ids`. The requirement is that when we re-parse and plan `persisted-sql` we should end up with the exact same set of `GlobalId`s that are in `resolved-ids`. That's (probably) why the catalog validations were failing in testdrive.

Reply:
Ah, that makes sense! We should write that down as a comment in the code.

Reply:
https://github.com/MaterializeInc/materialize/pull/24016